include/envoy/registry/registry.h

```diff
@@ -257,6 +297,20 @@ template <class T, class Base> class RegisterFactory {
     }
   }
 
+  /**
+   * Constructor that registers an instance of the factory with the FactoryRegistry along with
+   * vendor specific version.
+   */
+  explicit RegisterFactory(const envoy::api::v2::core::SemanticVersion& version) {
+    if (!instance_.name().empty()) {
+      FactoryRegistry<Base>::registerFactory(instance_, instance_.name(), version);
+    }
+
+    if (!FactoryCategoryRegistry::isRegistered(Base::category())) {
+      FactoryCategoryRegistry::registerCategory(Base::category(), new FactoryCategoryNames<Base>());
+    }
+  }
```

Comment:
@jpeach WDYT?

Reply:
This probably should allow the caller to register factories with both deprecated names and versions.

I'm a little hesitant about using the protobuf in the API; it seems awkward and we could easily provide a convenience wrapper.

Suggested syntax:
```C
REGISTER_FACTORY(FactoryBase, NamedFactory, FACTORY_VERSION(1, 2, 3)){"deprecated", "names"};
```

Alternatively, require a virtual `version()` method that factories can override (could introduce a required factory base type):
```C
struct Factory {
    virtual string_view name() const = 0;
    virtual optional<SemanticVersion> version() const { return nullopt; }
};
```

Any chance you'd like to rebase this onto #9125?

Reply:
I decided against the base class approach for now, given that so far we'd have very few versioned factories and a lot of unversioned. I can change my mind though if there are other compelling reasons to introduce base class for factories.